fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexFunctionHelper.java

```diff
@@ -159,15 +160,33 @@ public static KeyExpression getGroupedKey(@Nonnull KeyExpression key) {
             return key;
         }
         GroupingKeyExpression grouping = (GroupingKeyExpression) key;
-        return getSubKey(grouping.getWholeKey(), grouping.getGroupingCount(), grouping.getColumnSize());
+        return groupKey(getSubKey(grouping.getWholeKey(), grouping.getGroupingCount(), grouping.getColumnSize()));
     }
 
     public static KeyExpression getGroupingKey(@Nonnull KeyExpression key) {
         if (!(key instanceof GroupingKeyExpression)) {
             return EmptyKeyExpression.EMPTY;
         }
         GroupingKeyExpression grouping = (GroupingKeyExpression) key;
-        return getSubKey(grouping.getWholeKey(), 0, grouping.getGroupingCount());
+        return groupKey(getSubKey(grouping.getWholeKey(), 0, grouping.getGroupingCount()));
+    }
+
+    protected static KeyExpression groupKey(@Nonnull KeyExpression key) {
+        if (key instanceof ThenKeyExpression) {
```

Comment:
It seems like, at least in principle, that this should also recurse down, say, nesting key expressions and remove `Empty` expressions children... On the other hand, that might be better included in some kind of "key canonicalization" function that recurses down all of the various key types.

I could also see the argument that we should fix this to make "incremental progress" and then add full canonicalization later (not letting the perfect be the enemy of the good, etc.)

Reply:
I could see the benefit of a centralized canonical form transform. Though some care is required in plumbing that in, since key expressions are persisted in meta-data and things might rely on equality comparisons there.

But I was thinking of this as local to the these functions and therefore only aiming to take care of the ones for which it is responsible. I do understand that in some of the cases, the affected `Then` actually existed in the input and wasn't constructed here. But per the contract, it could have been, or, looked at another way, the result is now the same as if it fell into the general "splitting" logic.

So I think I would prefer to take this rather than aiming for the more comprehensive one.

Reply:
Yeah, I think that makes sense